src/tools/chip-cert/CertUtils.cpp

```diff
@@ -0,0 +1,826 @@
+    }
+
+    // Set the certificate version (must be 2, a.k.a. v3).
+    if (!X509_set_version(newCert, 2))
+    {
+        ReportOpenSSLErrorAndExit("X509_set_version", res = false);
+    }
+
+    // Generate a serial number for the cert.
+    res = SetCertSerialNumber(newCert);
+    VerifyTrueOrExit(res);
+
+    // Set the certificate validity time.
+    res = SetValidityTime(newCert, validFrom, validDays);
+    VerifyTrueOrExit(res);
+
+    // Set the certificate's public key.
+    if (!X509_set_pubkey(newCert, newKey))
+    {
+        ReportOpenSSLErrorAndExit("X509_set_pubkey", res = false);
+    }
+
+    // Set certificate subject DN.
+    res = subjectDN->SetCertSubjectDN(newCert);
+    VerifyTrueOrExit(res);
+
+    // Set the issuer name for the certificate. In the case of a self-signed cert, this will be
+    // the new cert's subject name.
+    if (!X509_set_issuer_name(newCert, X509_get_subject_name(caCert)))
+    {
+        ReportOpenSSLErrorAndExit("X509_set_issuer_name", res = false);
+    }
+
+    // Add the appropriate certificate extensions.
+    if (certType == kCertType_Node)
+    {
+        res = AddExtension(newCert, NID_basic_constraints, "critical,CA:FALSE") &&
+            AddExtension(newCert, NID_key_usage, "critical,digitalSignature,keyEncipherment") &&
+            AddExtension(newCert, NID_ext_key_usage, "critical,clientAuth,serverAuth");
+    }
+    else if (certType == kCertType_FirmwareSigning)
+    {
+        res = AddExtension(newCert, NID_basic_constraints, "critical,CA:FALSE") &&
+            AddExtension(newCert, NID_key_usage, "critical,digitalSignature") &&
+            AddExtension(newCert, NID_ext_key_usage, "critical,codeSigning");
+    }
+    else if (certType == kCertType_ICA || certType == kCertType_Root)
+    {
+        res = AddExtension(newCert, NID_basic_constraints, "critical,CA:TRUE") &&
+            AddExtension(newCert, NID_key_usage, "critical,keyCertSign,cRLSign");
+    }
+    VerifyTrueOrExit(res);
+
+    // Add a subject key id extension for the certificate.
+    res = AddSubjectKeyId(newCert);
+    VerifyTrueOrExit(res);
+
+    // Add the authority key id extension from the signing certificate. For self-signed cert's this will
+    // be the same as new cert's subject key id extension.
+    res = AddAuthorityKeyId(newCert, caCert);
+    VerifyTrueOrExit(res);
+
+    if (extSub != nullptr)
+    {
+        res = AddExtension(newCert, NID_subject_alt_name, extSub);
+        VerifyTrueOrExit(res);
+    }
+
+    if (extInfo != nullptr)
+    {
+        res = AddExtension(newCert, NID_info_access, extInfo);
+        VerifyTrueOrExit(res);
+    }
+
+    // Sign the new certificate.
+    if (!X509_sign(newCert, caKey, EVP_sha256()))
+    {
+        ReportOpenSSLErrorAndExit("X509_sign", res = false);
+    }
+
+exit:
+    if (newCert != nullptr && !res)
+    {
+        X509_free(newCert);
+    }
+    if (!keySupplied && newKey != nullptr && !res)
+    {
+        EVP_PKEY_free(newKey);
```

Comment:
Can you use std::unique_ptr with a custom deleter here?

Reply:
done